python/pyarrow/serialization.pxi

```diff
@@ -0,0 +1,233 @@
+            deserialize objects of the class in a particular way.
+    """
+    type_to_type_id[type] = type_id
+    whitelisted_types[type_id] = type
+    if pickle:
+        types_to_pickle.add(type_id)
+    if custom_serializer is not None:
+        custom_serializers[type_id] = custom_serializer
+        custom_deserializers[type_id] = custom_deserializer
+
+def serialization_callback(obj):
+    if type(obj) not in type_to_type_id:
+        raise SerializationException("pyarrow does not know how to "
+                                     "serialize objects of type {}."
+                                     .format(type(obj)),
+                                     obj)
+    type_id = type_to_type_id[type(obj)]
+    if type_id in types_to_pickle:
+        serialized_obj = {"data": pickle.dumps(obj), "pickle": True}
+    elif type_id in custom_serializers:
+        serialized_obj = {"data": custom_serializers[type_id](obj)}
+    else:
+        if is_named_tuple(type(obj)):
+            serialized_obj = {}
+            serialized_obj["_pa_getnewargs_"] = obj.__getnewargs__()
+        elif hasattr(obj, "__dict__"):
+            serialized_obj = obj.__dict__
+        else:
+            raise SerializationException("We do not know how to serialize "
+                                         "the object '{}'".format(obj), obj)
+    return dict(serialized_obj, **{"_pytype_": type_id})
+
+def deserialization_callback(serialized_obj):
+    type_id = serialized_obj["_pytype_"]
+
+    if "pickle" in serialized_obj:
+        # The object was pickled, so unpickle it.
+        obj = pickle.loads(serialized_obj["data"])
+    else:
+        assert type_id not in types_to_pickle
+        if type_id not in whitelisted_types:
+            raise "error"
+        type = whitelisted_types[type_id]
+        if type_id in custom_deserializers:
+            obj = custom_deserializers[type_id](serialized_obj["data"])
+        else:
+            # In this case, serialized_obj should just be the __dict__ field.
+            if "_pa_getnewargs_" in serialized_obj:
+                obj = type.__new__(type, *serialized_obj["_pa_getnewargs_"])
+            else:
+                obj = type.__new__(type)
+                serialized_obj.pop("_pytype_")
+                obj.__dict__.update(serialized_obj)
+    return obj
+
+pyarrow_serialize_callback = <PyObject*> serialization_callback
+
+pyarrow_deserialize_callback = <PyObject*> deserialization_callback
+
+# Main entry point for serialization
+def serialize_sequence(object value):
+    cdef int32_t recursion_depth = 0
+    cdef PythonObject result = PythonObject()
+    cdef c_vector[PyObject*] sequences
+    cdef shared_ptr[CArray] array
+    cdef c_vector[PyObject*] tensors
+    cdef PyObject* tensor
+    cdef shared_ptr[CTensor] out
+    sequences.push_back(<PyObject*> value)
+    check_status(SerializeSequences(sequences, recursion_depth, &array, tensors))
```

Comment:
Release GIL (reacquire within -- the `with gil` in check_status can cause deadlocks with subinterpreters)